File: src/main/java/de/rubixdev/enchantedshulkers/interfaces/EnchantableBlockEntity.java
package de.rubixdev.enchantedshulkers.interfaces;

import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtList;

public interface EnchantableBlockEntity {
    NbtList enchantedShulkers$getEnchantments();

    void enchantedShulkers$setEnchantments(NbtList enchantments);

    default NbtCompound enchantedShulkers$toClientNbt() {
        // Only send the enchantments to the client to reduce packet size
        NbtCompound nbt = new NbtCompound();
        nbt.put("Enchantments", enchantedShulkers$getEnchantments());
        return nbt;
    }
}

File: src/main/kotlin/de/rubixdev/enchantedshulkers/interfaces/EnchantableBlockEntity.kt
package de.rubixdev.enchantedshulkers.interfaces

import net.minecraft.nbt.NbtCompound

//#if MC >= 12006
import net.minecraft.block.entity.BlockEntity
import net.minecraft.component.type.ItemEnchantmentsComponent
import net.minecraft.nbt.NbtOps
//#else
//$$ import net.minecraft.nbt.NbtList
//#endif

@Suppress("FunctionName")
interface EnchantableBlockEntity {
    //#if MC >= 12006
    fun `enchantedShulkers$getEnchantments`(): ItemEnchantmentsComponent

    fun `enchantedShulkers$setEnchantments`(enchantments: ItemEnchantmentsComponent)
    //#else
    //$$ fun `enchantedShulkers$getEnchantments`(): NbtList

    //$$ fun `enchantedShulkers$setEnchantments`(enchantments: NbtList)
    //#endif

    fun `enchantedShulkers$toClientNbt`() =
        //#if MC >= 12006
        when (this) {
            is BlockEntity -> BlockEntity.Components.CODEC.encodeStart(NbtOps.INSTANCE, components).getOrThrow() as NbtCompound
            else -> NbtCompound()
        }
        //#else
        //$$ NbtCompound().apply {
        //$$     put("Enchantments", `enchantedShulkers$getEnchantments`())
        //$$ }
        //#endif
}
